table/builder.go

```diff
@@ -177,30 +191,20 @@ func (b *Builder) Add(key []byte, value y.ValueStruct) error {
 
 // ReachedCapacity returns true if we... roughly (?) reached capacity?
 func (b *Builder) ReachedCapacity(cap int64) bool {
-	estimateSz := b.buf.Len() + 8 /* empty header */ + 4*len(b.restarts) + 8 // 8 = end of buf offset + len(restarts).
+	estimateSz := b.buf.Len() + 8 /* empty header */ + 4 /* Index length */ +
+		5*(len(b.tableIndex.KeyOffSetList)) /* approximate index size */
 	return int64(estimateSz) > cap
 }
 
 // blockIndex generates the block index for the table.
-// It is mainly a list of all the block base offsets.
 func (b *Builder) blockIndex() []byte {
-	// Store the end offset, so we know the length of the final block.
-	b.restarts = append(b.restarts, uint32(b.buf.Len()))
-
-	// Add 4 because we want to write out number of restarts at the end.
-	sz := 4*len(b.restarts) + 4
-	out := make([]byte, sz)
-	buf := out
-	for _, r := range b.restarts {
-		binary.BigEndian.PutUint32(buf[:4], r)
-		buf = buf[4:]
-	}
-	binary.BigEndian.PutUint32(buf[:4], uint32(len(b.restarts)))
+	out, err := b.tableIndex.Marshal()
+	y.Check(err)
 	return out
 }
 
 // Finish finishes the table by appending the index.
-func (b *Builder) Finish() []byte {
+func (b *Builder) Finish() ([]byte, error) {
```

Comment:
Are we returning an error at all anywhere in this function?

Reply:
Yes, the checksum validation returns error on failure.